### PR TITLE
Building api

### PR DESCRIPTION
**Pull request by felipe74590**
creating api

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -6,10 +6,12 @@ readme = "README.md"
 requires-python = ">=3.12"
 dependencies = [
     "beautifulsoup4>=4.12.3",
+    "fastapi>=0.115.0",
```
Comment (bbelderbos) [inline pyproject.toml]: As discuss today and I think you already did:

```
uv add "fastapi[standard]"
uv run fastapi dev src/api.py
```

File: src/api.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+def delete_event(event_id: int):
+    with Session(engine) as session:
+        event = session.get(Event, event_id)
+        if not event:
+            raise HTTPException(status_code=404, detail="Event not found")
+        session.delete(event)
+        session.commit()
+    return event
+
+if __name__ == "__main__":
```
Comment (bbelderbos) [inline src/api.py]: I think you can ditch this with the new fastapi cli way of running the app

File: src/api.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+app = FastAPI()
+
+DATABASE_URL = config("DATABASE")
+engine = create_engine(DATABASE_URL)
+
+SQLModel.metadata.create_all(engine)
+
+@app.get("/events/", response_model=list[Event])
+def read_events(skip:int=0, limit: int = 20):
```
Comment (bbelderbos) [inline src/api.py]: would be cool to make an infinite scroll if you want to build a web interface, I did a video on this a while ago using htmx:

[video] How to make an infinite scroll of YouTube videos using FastAPI, SQLModel and htmx
https://www.youtube.com/watch?v=5uOCUkJU-4Q

File: src/api.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+            raise HTTPException(status_code=404, detail="Event not found")
+    return event
+
+@app.put("/events/{event_id}", response_model=Event)
+def update_event(event_id: int, event: Event):
+    with Session(engine) as session:
+        db_event = session.get(Event, event_id)
+        if not db_event:
+            raise HTTPException(status_code=404, detail="Event not found")
+        event.id = event_id
```
Comment (bbelderbos) [inline src/api.py]: seems you need to put back the new event:

`db_event.sqlmodel_update(event)` 

(`event: eventUpdate` to be defined/added here)

TIL:

image

Comment (bbelderbos) [reply src/api.py]: https://sqlmodel.tiangolo.com/tutorial/fastapi/update/#update-the-hero-in-the-database

**Review by bbelderbos — APPROVED** [review]
Good work, keep it up!

File: src/data/db_helper.py [hunk 5]
```diff
@@ -59,16 +61,12 @@ async def migrate_to_postgresql():
 			existing_event = await session.execute(select(Event).where(Event.id == row[0]))
 			if not existing_event:
 				event_instance = Event(
-					id=row[0],
-					title=row[1],
-					venue=row[2],
-					start_datetime=row[3],
-					category=row[4],
-					event_link=row[5]
+					id=row[0], title=row[1], venue=row[2], start_datetime=row[3], category=row[4], event_link=row[5]
```
Comment (bbelderbos) [inline src/data/db_helper.py]: use `*row`?

File: src/data/db_helper.py [hunk 6]
```diff
@@ -38,16 +37,19 @@ def add_events_to_db(events):
...
 		session.commit()
 
+
 postgresql_url = config("PSQL_MIRGRATE_URL")
+
+
 async def migrate_to_postgresql():
 	"""Migrate data in sqlite databse atx_events to postgreSQL database atx-events"""
-	sqlite_conn = sqlite3.connect(sqlite_db)
+	sqlite_conn = sqlite3.connect(sqlite_db_name)
```
Comment (bbelderbos) [inline src/data/db_helper.py]: do we need this? we don't have a production database yet, when we do I think we just start with an empty db, same for switching the local DB from sqlite to postgres

File: src/api/main_api.py [hunk 7]
```diff
@@ -1,20 +1,45 @@
...
 SQLModel.metadata.create_all(engine)
 
 
 @app.get("/events/", response_model=list[Event])
-def read_events(skip: int = 0, limit: int = 20):
+def read_events(
+	skip: int = 0,
+	limit: int = 50,
+	from_date: Optional[datetime] = None,
+	to_date: Optional[datetime] = None,
```
Comment (bbelderbos) [inline src/api/main_api.py]: you can now also use `datetime | None = None ` (since 3.9 I believe)

File: src/api/main_api.py [hunk 8]
```diff
@@ -1,20 +1,45 @@
...
+	Get a list of events.
+	- **limit**: Maximum number of records to return.
+	- **from_date**: Filter events starting on this date
+	- **to_date**: Filter events till this date
+	- **venue_keyword**: Search for events in the venue name.
+	"""
 	with Session(engine) as session:
+		statement = select(Event)
+
+		if from_date:
```
Comment (bbelderbos) [inline src/api/main_api.py]: `is not None` - I usually do it like this when a default value is `None`

File: src/api/main_api.py [hunk 9]
```diff
@@ -1,20 +1,45 @@
...
+			statement = statement.where(Event.start_datetime >= from_date)
+		if to_date:
```
Comment (bbelderbos) [inline src/api/main_api.py]: nice

**Review by bbelderbos — APPROVED** [review]
Cool, let me try it. Let's discuss the data migration

**Comment by felipe74590** [comment]
Working on solving https://github.com/felipe74590/atx-events/issues/25#issuecomment-2387564294 before finishing migrations of updates.

File: src/api.py (hunk 3 above)
Comment (felipe74590) [reply src/api.py]: Will add this when working on UI.

**Pull request merged**